src/c_api.cc

```diff
@@ -120,8 +120,45 @@ inline void Allreduce(void *sendrecvbuf,
     default: utils::Error("unknown enum_op");
   }
 }
-
-
+inline void AllgatherRing(void *sendrecvbuf_,
```

Comment:
Why do you need inline for this?  It might be some legacy convention floating around with it, but those are not necessary.

Reply:
Honestly, I've chosen to define this function as inline because similar functions but for allreduce are defined as inline. This function is very simple (computationally) and has only one caller. Defining this function as inline won't do any harm. It would avoid context switching between the caller and this function if the compiler agrees to inline this function. But if you prefer, I can define it as a regular function.

Reply:
@nateagr  inline is a specifier for linker not for compiler despite its name.  See https://stackoverflow.com/questions/1759300/when-should-i-write-the-keyword-inline-for-a-function-method  It doesn't do any harm but just not necessary in 2019.  I won't block the PR for this small detail as you put a lots of effort and It's really useful changes.  But it's also not harmful to clean up the code a little bit. ;-)

Reply:
Ok I'll do the cleanup in c_api.cc then. Thanks for the clarification.